File: fetch-api/fetch_api/src/health_checker.py
from datetime import (datetime, timedelta)
from fetch_api.src.telemetry.logging import log
from fetch_api.settings import (settings, connectors)
from fetch_api.src.client import ConnectorClient

from common.telemetry.src.tracing.wrappers import traced
from common.telemetry.src.tracing.helpers import reword



class HealthChecker:
    def __init__(self, scheduler, connector):
        self.scheduler = scheduler
        self.connector = connector


    def get_next_interval(self):
        return (
            settings.connector_health_check_interval_seconds
            if self.connector.healthy is True
            else settings.connector_health_retry_interval_seconds
        )


    def get_next_check_time(self):
        ts = datetime.now() + timedelta(seconds=self.get_next_interval())
        return ts.isoformat().split('.')[0]


    @traced()
    def get_connector_status(self, span=None):
        was_healthy = self.connector.healthy
        self.connector.health_last_check = datetime.now().isoformat().split('.')[0]

        try:
            response = ConnectorClient.ping(
                connector_name=self.connector.name,
                connector_url=self.connector.url,
                health_endpoint=self.connector.health_endpoint
            )

        except Exception as err:
            self.connector.healthy = None

            if was_healthy is not None:
                self.update_connector_schedule()

            else:
                self.connector.health_next_check = self.get_next_check_time()

            log.critical(f'Health check failed, connector is unreachable', extra={
                'connector': self.connector.name,
                'health_endpoint': self.connector.health_endpoint,
                'error': str(err)
            })
            span.set_attributes(
                reword({
                    'health.connector.name': self.connector.name,
                    'health.connector.endpoint': self.connector.health_endpoint,
                    'health.connector.error.message': str(err),
                    'health.connector.error.type': type(err).__name__,
                    'health.connector.last_check': self.connector.health_last_check,
                    'health.connector.next_check': self.connector.health_next_check,
                    'health.connector.status.current': self.connector.healthy,
                    'health.connector.status.previous': was_healthy
                })
            )

            return None
        
        try:
            is_healthy = response.json()['healthy']

        except Exception as err:
            self.connector.healthy = None

            if was_healthy is not None:
                self.update_connector_schedule()

            else:
                self.connector.health_next_check = self.get_next_check_time()

            log.critical(f'Health check failed, got unexpected response', extra={
                'connector': self.connector.name,
                'health_endpoint': self.connector.health_endpoint,
                'error': str(err)
            })
            span.set_attributes(
                reword({
                    'health.connector.name': self.connector.name,
                    'health.connector.endpoint': self.connector.health_endpoint,
                    'health.connector.error.message': str(err),
                    'health.connector.error.type': type(err).__name__,
                    'health.connector.response.content': response.text,
                    'health.connector.response.status_code': response.status_code,
                    'health.connector.last_check': self.connector.health_last_check,
                    'health.connector.next_check': self.connector.health_next_check,
                    'health.connector.status.current': self.connector.healthy,
                    'health.connector.status.previous': was_healthy
                })
            )

            return None

        if isinstance(is_healthy, bool):
            if (response.status_code == 200) and (is_healthy is True):
                self.connector.healthy = True

                if was_healthy is not True:
                    log.debug('Health check successful', extra={
                        'connector': self.connector.name,
                        'health_endpoint': self.connector.health_endpoint
                    })
                    self.update_connector_schedule()

                else:
                    self.connector.health_next_check = self.get_next_check_time()
                
                span.set_attributes(
                    reword({
                        'health.connector.name': self.connector.name,
                        'health.connector.endpoint': self.connector.health_endpoint,
                        'health.connector.last_check': self.connector.health_last_check,
                        'health.connector.next_check': self.connector.health_next_check,
                        'health.connector.status.current': self.connector.healthy,
                        'health.connector.status.previous': was_healthy
                    })
                )

            else:
                self.connector.healthy = False

                if was_healthy is not False:
                    self.update_connector_schedule()

                else:
                    self.connector.health_next_check = self.get_next_check_time()
                
                log.critical('Health check failed', extra={
                    'connector': self.connector.name,
                    'health_endpoint': self.connector.health_endpoint,
                    'response_status_code': response.status_code
                })
                span.set_attributes(
                    reword({
                        'health.connector.name': self.connector.name,
                        'health.connector.endpoint': self.connector.health_endpoint,
                        'health.connector.response.content': response.text,
                        'health.connector.response.status_code': response.status_code,
                        'health.connector.last_check': self.connector.health_last_check,
                        'health.connector.next_check': self.connector.health_next_check,
                        'health.connector.status.current': self.connector.healthy,
                        'health.connector.status.previous': was_healthy
                    })
                )

        elif is_healthy is None:
            self.connector.healthy = None

            if was_healthy is not None:
                self.update_connector_schedule()

            else:
                self.connector.health_next_check = self.get_next_check_time()

            log.critical(f'Health check failed, got unexpected response', extra={
                'connector': self.connector.name,
                'health_endpoint': self.connector.health_endpoint,
                'response_status_code': response.status_code
            })
            span.set_attributes(
                reword({
                    'health.connector.name': self.connector.name,
                    'health.connector.endpoint': self.connector.health_endpoint,
                    'health.connector.response.content': response.text,
                    'health.connector.response.status_code': response.status_code,
                    'health.connector.last_check': self.connector.health_last_check,
                    'health.connector.next_check': self.connector.health_next_check,
                    'health.connector.status.current': self.connector.healthy,
                    'health.connector.status.previous': was_healthy
                })
            )


    @traced()
    def create_connector_schedule(self, span=None):
        log.debug(f'Scheduling health checks for {self.connector.name}')
        self.get_connector_status()

        if self.connector.health_job_id is None:
            self.update_connector_schedule()


    @traced()
    def update_connector_schedule(self, span=None):
        if not self.connector.health_job_id is None:
            try:
                self.scheduler.remove_job(self.connector.health_job_id)

            except:
                pass

        interval_seconds = self.get_next_interval()
        self.connector.health_next_check = self.get_next_check_time()

        job = self.scheduler.add_job(
            self.get_connector_status,
            'interval',
            seconds=interval_seconds,
            id=f'health_check_{self.connector.name}'
        )

        self.connector.health_job_id = job.id
        
        span.set_attributes(
            reword({
                'connector.name': self.connector.name,
                'connector.scheduler.interval.seconds': interval_seconds,
                'connector.scheduler.job.id': self.connector.health_job_id,
                'health.connector.last_check': self.connector.health_last_check,
                'health.connector.next_check': self.connector.health_next_check
            })
        )

File: fetch-api/connectors/slack/internal/health/health_checker.go
package health

import (
	"time"

	"connector-slack/internal/config"
	"connector-slack/internal/slack"
	t "connector-slack/internal/telemetry"

	"github.com/go-co-op/gocron"
)

type HealthChecker struct {
	Scheduler *gocron.Scheduler
}

func (instance *HealthChecker) CreateSchedule() {
	t.Log.Info("Scheduling health checks..")
	instance.updateStatus()

	if config.Config.HealthJobID == nil {
		instance.updateSchedule()
	}
}

func (instance *HealthChecker) getNextInterval() int {
	if config.Config.Healthy != nil && *config.Config.Healthy {
		return config.Config.HealthCheckIntervalSeconds
	} else {
		return config.Config.HealthRetryIntervalSeconds
	}
}

func (instance *HealthChecker) getNextCheckTime() string {
	ts := time.Now().Add(
		time.Duration(
			instance.getNextInterval(),
		) * time.Second,
	)

	return ts.Format("2006-01-02T15:04:05")
}

func (instance *HealthChecker) updateStatus() {
	var wasHealthy bool
	var isHealthy bool

	wasHealthyPtr := config.Config.Healthy

	if wasHealthyPtr != nil {
		wasHealthy = *wasHealthyPtr
	}

	lastCheckTime := time.Now().Format("2006-01-02T15:04:05")
	config.Config.HealthLastCheck = &lastCheckTime
	err := slack.Client.Ping()

	if err != nil {
		config.Config.Healthy = nil

		if wasHealthyPtr != nil {
			instance.updateSchedule()
		} else {
			nextCheckTime := instance.getNextCheckTime()
			config.Config.HealthNextCheck = &nextCheckTime
		}

		t.Log.Error("Health check failed, Slack is unreachable", "error", err.Error())

		return
	}

	isHealthy = true
	config.Config.Healthy = &isHealthy

	if wasHealthyPtr != nil && wasHealthy {
		nextCheckTime := instance.getNextCheckTime()
		config.Config.HealthNextCheck = &nextCheckTime
	} else {
		t.Log.Info("Health check successful")
		instance.updateSchedule()
	}
}

func (instance *HealthChecker) updateSchedule() {
	if instance.Scheduler == nil {
		return
	}

	if config.Config.HealthJobID != nil {
		instance.Scheduler.RemoveByTag(*config.Config.HealthJobID)
	}

	interval := instance.getNextInterval()
	nextCheckTime := instance.getNextCheckTime()
	config.Config.HealthNextCheck = &nextCheckTime

	jobTag := "health_check_slack"
	job, _ := instance.Scheduler.Every(interval).Seconds().Do(instance.updateStatus)
	job.Tag(jobTag)
	config.Config.HealthJobID = &jobTag
}
